File: backend/app/routers/goals.py
import json

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from app.database import get_db
from app.schemas import (
    GoalCreate,
    GoalResponse,
    GoalUpdate,
    KeyResultCreate,
    KeyResultResponse,
    KeyResultUpdate,
)
from app.services import goal_service

router = APIRouter(prefix="/goals", tags=["goals"])


def _kr_response(db: Session, kr, goal) -> KeyResultResponse:
    current, progress = goal_service.kr_progress(db, kr, goal)
    try:
        link = json.loads(kr.link or "{}")
    except (TypeError, json.JSONDecodeError):
        link = {}
    return KeyResultResponse(
        id=kr.id,
        goal_id=kr.goal_id,
        title=kr.title,
        kind=kr.kind,
        target_value=kr.target_value,
        current_value=current,
        unit=kr.unit,
        link=link if isinstance(link, dict) else {},
        progress=progress,
    )


def _goal_response(db: Session, goal) -> GoalResponse:
    return GoalResponse(
        id=goal.id,
        title=goal.title,
        notes=goal.notes,
        status=goal.status,
        start_date=goal.start_date,
        end_date=goal.end_date,
        sort_order=goal.sort_order,
        progress=goal_service.goal_progress(db, goal),
        key_results=[_kr_response(db, kr, goal) for kr in goal.key_results],
        created_at=goal.created_at,
    )


@router.post("", response_model=GoalResponse, status_code=status.HTTP_201_CREATED)
def create_goal(payload: GoalCreate, db: Session = Depends(get_db)):
    return _goal_response(db, goal_service.create_goal(db, payload))


@router.get("", response_model=list[GoalResponse])
def list_goals(
    include_archived: bool = Query(False), db: Session = Depends(get_db)
):
    return [
        _goal_response(db, g) for g in goal_service.list_goals(db, include_archived)
    ]


@router.get("/{goal_id}", response_model=GoalResponse)
def get_goal(goal_id: int, db: Session = Depends(get_db)):
    goal = goal_service.get_goal(db, goal_id)
    if goal is None:
        raise HTTPException(status_code=404, detail="目标不存在")
    return _goal_response(db, goal)


@router.put("/{goal_id}", response_model=GoalResponse)
def update_goal(goal_id: int, payload: GoalUpdate, db: Session = Depends(get_db)):
    goal = goal_service.update_goal(db, goal_id, payload)
    if goal is None:
        raise HTTPException(status_code=404, detail="目标不存在")
    return _goal_response(db, goal)


@router.delete("/{goal_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_goal(goal_id: int, db: Session = Depends(get_db)):
    if not goal_service.soft_delete_goal(db, goal_id):
        raise HTTPException(status_code=404, detail="目标不存在")


@router.post("/{goal_id}/krs", response_model=KeyResultResponse, status_code=status.HTTP_201_CREATED)
def add_key_result(goal_id: int, payload: KeyResultCreate, db: Session = Depends(get_db)):
    kr = goal_service.add_key_result(db, goal_id, payload)
    if kr is None:
        raise HTTPException(status_code=404, detail="目标不存在")
    goal = goal_service.get_goal(db, goal_id)
    return _kr_response(db, kr, goal)


@router.put("/krs/{kr_id}", response_model=KeyResultResponse)
def update_key_result(kr_id: int, payload: KeyResultUpdate, db: Session = Depends(get_db)):
    kr = goal_service.update_key_result(db, kr_id, payload)
    if kr is None:
        raise HTTPException(status_code=404, detail="关键结果不存在")
    goal = goal_service.get_goal(db, kr.goal_id)
    return _kr_response(db, kr, goal)


@router.delete("/krs/{kr_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_key_result(kr_id: int, db: Session = Depends(get_db)):
    if not goal_service.delete_key_result(db, kr_id):
        raise HTTPException(status_code=404, detail="关键结果不存在")

File: frontend/src/api/goals.js
// OKR 目标 API 封装。相对路径 /goals：dev 经 Vite proxy，prod 同源走后端
// KR 的 link 约定：manual → {}；tag_task_count → { tag: "标签名" }；habit_checkins → { habit_id: 数字 }
const BASE = '/goals'

async function parse(res) {
  if (!res.ok) {
    const text = await res.text().catch(() => '')
    throw new Error(`请求失败 (${res.status}) ${text}`)
  }
  if (res.status === 204) return null
  return res.json()
}

// 默认带上已归档目标（界面置灰展示，便于取消归档）；后端已按 sort_order 排序
export function listGoals() {
  return fetch(`${BASE}?include_archived=true`).then(parse)
}

export function createGoal(goal) {
  return fetch(BASE, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(goal),
  }).then(parse)
}

export function updateGoal(id, patch) {
  return fetch(`${BASE}/${id}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(patch),
  }).then(parse)
}

export function deleteGoal(id) {
  return fetch(`${BASE}/${id}`, { method: 'DELETE' }).then(parse)
}

export function createKeyResult(goalId, kr) {
  return fetch(`${BASE}/${goalId}/krs`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(kr),
  }).then(parse)
}

// patch 可含 title/kind/target_value/unit/link/current_value（current_value 仅 manual 生效）
export function updateKeyResult(krId, patch) {
  return fetch(`${BASE}/krs/${krId}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(patch),
  }).then(parse)
}

export function deleteKeyResult(krId) {
  return fetch(`${BASE}/krs/${krId}`, { method: 'DELETE' }).then(parse)
}
